src/Hosting/Hosting/src/Internal/HostingMetrics.cs

```diff
@@ -0,0 +1,174 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Diagnostics;
+using System.Diagnostics.Metrics;
+using Microsoft.AspNetCore.Http;
+using Microsoft.Extensions.Metrics;
+
+namespace Microsoft.AspNetCore.Hosting;
+
+internal sealed class HostingMetrics : IDisposable
+{
+    public const string MeterName = "Microsoft.AspNetCore.Hosting";
+
+    private readonly Meter _meter;
+    private readonly UpDownCounter<long> _currentRequestsCounter;
+    private readonly Histogram<double> _requestDuration;
+
+    public HostingMetrics(IMeterFactory meterFactory)
+    {
+        _meter = meterFactory.CreateMeter(MeterName);
+
+        _currentRequestsCounter = _meter.CreateUpDownCounter<long>(
+            "current-requests",
+            description: "Current requests");
+
+        _requestDuration = _meter.CreateHistogram<double>(
+            "request-duration",
+            description: "Request duration");
+    }
+
+    public void RequestStart(string scheme, string method, HostString host)
+    {
+        // Tags must match request end.
+        var tags = CreateRequestTags(scheme, method, host);
+        _currentRequestsCounter.Add(1, tags);
+    }
+
+    public void RequestEnd(string scheme, string method, HostString host, string? routePattern, int statusCode, Exception? exception, IList<KeyValuePair<string, object?>>? customTags, long startTimestamp, long currentTimestamp)
+    {
+        // Tags must match request start.
+        var tags = CreateRequestTags(scheme, method, host);
+        _currentRequestsCounter.Add(-1, tags);
+
+        // Add information gathered during request.
+        tags.Add("status-code", GetBoxedStatusCode(statusCode));
+        if (routePattern != null)
+        {
+            tags.Add("route", routePattern);
+        }
+        // This exception is only present if there is an unhandled exception.
+        // Exceptions caught by ExceptionHandlerMiddleware and DeveloperExceptionMiddleware add exception-name to custom tags.
+        if (exception != null)
+        {
+            tags.Add("exception-name", exception.GetType().Name);
+        }
+        if (customTags != null)
+        {
+            for (var i = 0; i < customTags.Count; i++)
+            {
+                tags.Add(customTags[i]);
+            }
+        }
+
+        var duration = Stopwatch.GetElapsedTime(startTimestamp, currentTimestamp);
+        _requestDuration.Record(duration.TotalMilliseconds, tags);
+    }
+
+    public void Dispose()
+    {
+        _meter.Dispose();
+    }
+
+    public bool IsEnabled() => _currentRequestsCounter.Enabled || _requestDuration.Enabled;
+
+    private static TagList CreateRequestTags(string scheme, string method, HostString host)
+    {
+        var tags = new TagList();
+        tags.Add("scheme", scheme);
```

Comment:
is it possible to have some extensibility to customize these tag names and potentially allowing more to be added? 

The metric instrumentation library in OTel is emitting the same metric (_requestDuration) but it needs to follow the semantic conventions defined [here](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/semantic_conventions/http-metrics.md).

With extensibility we could avoid a separate instrumentation library. 

@reyang @noahfalk @CodeBlanch not sure if this was already discussed.

Reply:
There isn't a way to customize tag names here. However, it wouldn't be difficult to have a `MeterListener` that listens to a counter, get tag values by their .NET name, and then report a new counter with tag names that match OTel.

Reply:
My hope is that such processing/enrichment can be done in a generic way (e.g. in OpenTelemetry .NET SDK or in .NET BCL DiagnosticSource lib) rather than having ASP.NET Core or any other libraries doing it in their own/different ways - unless there is any critical reason to go against this direction (e.g. extreme perf, the ability to get more stuff from domain specific context, etc.).

Reply:
This PR adds low-level APIs. There isn't anything stopping ASP.NET Core, or a third-party library, from adding something on top that gathers tags and then sets them on the `IHttpMetricsTagsFeature`.